Content.Server/Implants/SubdermalImplantSystem.cs

```diff
@@ -107,41 +110,92 @@ private void OnScramImplant(EntityUid uid, SubdermalImplantComponent component,
+        // In reality the probability is higher and depends on the number of grids.
+        if (userXform.GridUid != null && TryComp<MapGridComponent>(userXform.GridUid, out var gridComp))
+        {
+            var userGrid = new Entity<MapGridComponent>(userXform.GridUid.Value, gridComp);
+            if (_random.Prob(0.5f))
             {
-                if (!_physicsQuery.TryGetComponent(entity, out var body))
-                    continue;
+                _targetGrids.Remove(userGrid);
+                targetGrid = userGrid;
+            }
+        }
 
-                if (body.BodyType != BodyType.Static ||
-                    !body.Hard ||
-                    (body.CollisionLayer & (int) CollisionGroup.Impassable) == 0)
-                    continue;
+        if (targetGrid == null)
+            targetGrid = _random.GetRandom().PickAndTake(_targetGrids);
 
-                valid = false;
-                break;
+        EntityCoordinates? targetCoords = null;
+
+        do
+        {
+            var valid = false;
+
+            var range = (float) Math.Sqrt(radius);
+            var box = Box2.CenteredAround(userCoords.Position, new Vector2(range, range));
+            var tilesInRange = _mapSystem.GetTilesEnumerator(targetGrid.Value.Owner, targetGrid.Value.Comp, box, false);
+            var tileList = new ValueList<Vector2i>();
+
+            while (tilesInRange.MoveNext(out var tile))
+            {
+                tileList.Add(tile.GridIndices);
+            }
+
+            while (tileList.Count != 0)
+            {
+                var tile = tileList.RemoveSwap(_random.Next(tileList.Count));
+                valid = true;
+                foreach (var entity in _mapSystem.GetAnchoredEntities(targetGrid.Value.Owner, targetGrid.Value.Comp,
+                             tile))
+                {
+                    if (!_physicsQuery.TryGetComponent(entity, out var body))
+                        continue;
+
+                    if (body.BodyType != BodyType.Static ||
+                        !body.Hard ||
+                        (body.CollisionLayer & (int) CollisionGroup.MobMask) == 0)
+                        continue;
+
+                    valid = false;
+                    break;
+                }
+
+                if (valid)
+                {
+                    targetCoords = new EntityCoordinates(targetGrid.Value.Owner,
+                        _mapSystem.TileCenterToVector(targetGrid.Value, tile));
+                    break;
+                }
             }
-            if (valid)
+
+            if (valid || _targetGrids.Count == 0) // if we don't do the check here then PickAndTake will blow up on an empty set.
                 break;
-        }
-        _xform.SetWorldPosition(ent, targetCoords.Position);
-        _audio.PlayPvs(implant.TeleportSound, ent);
 
-        args.Handled = true;
+            targetGrid = _random.GetRandom().PickAndTake(_targetGrids);
+        } while (true);
```

Comment:
Can't you just do while _targetGrids.Count > 0, or just do a regular while loop instead?

Reply:
Maybe if a rearranged some stuff in the function. At the time I made it I probably thought about it a bit weird. Do you want me to make a new PR with that?